File: sdks/rust/src/resources/invoices.rs
//! Invoice resource (`/v1/invoices`).

use std::collections::BTreeMap;

use reqwest::Method;
use serde::Serialize;

use crate::{Client, Invoice, Result};

/// Accessor for invoice operations.
#[derive(Debug)]
pub struct Invoices<'a> {
    client: &'a Client,
}

/// A line item for invoice creation.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateInvoiceLineItem {
    pub description: String,
    pub quantity: i64,
    pub unit_amount: String,
}

/// A tax rate applied to an invoice.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TaxRate {
    pub jurisdiction: String,
    pub rate_bps: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub inclusive: Option<bool>,
}

/// Body for creating a draft invoice (`POST /v1/invoices`).
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateInvoice {
    pub organization_id: String,
    pub customer_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line_items: Option<Vec<CreateInvoiceLineItem>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub discount: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tax_rate: Option<TaxRate>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub due_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<BTreeMap<String, String>>,
}

impl<'a> Invoices<'a> {
    pub(crate) fn new(client: &'a Client) -> Self {
        Invoices { client }
    }

    /// Create a draft invoice.
    pub async fn create(&self, body: &CreateInvoice) -> Result<Invoice> {
        self.client
            .request(Method::POST, "/v1/invoices", Some(body))
            .await
    }

    /// List invoices, optionally filtered by customer.
    pub async fn list(&self, customer_id: Option<&str>) -> Result<Vec<Invoice>> {
        let path = match customer_id {
            Some(cid) => format!("/v1/invoices?customerId={}", encode(cid)),
            None => "/v1/invoices".to_string(),
        };
        self.client.request_no_body(Method::GET, &path).await
    }

    /// Fetch an invoice by id.
    pub async fn get(&self, id: &str) -> Result<Invoice> {
        self.client
            .request_no_body(Method::GET, &format!("/v1/invoices/{id}"))
            .await
    }

    /// Transition a draft invoice to `open`.
    pub async fn finalize(&self, id: &str) -> Result<Invoice> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/invoices/{id}/finalize"), None)
            .await
    }

    /// Mark an open invoice as paid.
    pub async fn pay(&self, id: &str) -> Result<Invoice> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/invoices/{id}/pay"), None)
            .await
    }

    /// Void a draft or open invoice.
    pub async fn void(&self, id: &str) -> Result<Invoice> {
        self.client
            .request::<_, ()>(Method::POST, &format!("/v1/invoices/{id}/void"), None)
            .await
    }
}

/// Minimal percent-encoding for query-string values.
fn encode(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for byte in value.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(byte as char);
            }
            other => out.push_str(&format!("%{other:02X}")),
        }
    }
    out
}

File: sdks/python/settlekit/resources/invoices.py
"""Invoices resource — draft, finalize, pay, and void invoices."""

from __future__ import annotations

from typing import Any, Iterable, Mapping

from ._base import AsyncResource, SyncResource, clean_body


def _normalize_line_items(
    line_items: Iterable[Mapping[str, Any]]
) -> list[dict[str, Any]]:
    normalized: list[dict[str, Any]] = []
    for item in line_items:
        normalized.append(
            clean_body(
                {
                    "description": item.get("description"),
                    "quantity": item.get("quantity"),
                    "unitAmount": item.get("unitAmount") or item.get("unit_amount"),
                }
            )
        )
    return normalized


class Invoices(SyncResource):
    """Synchronous invoices namespace."""

    def create(
        self,
        *,
        organization_id: str,
        customer_id: str,
        line_items: Iterable[Mapping[str, Any]],
    ) -> Any:
        """Create a draft invoice. ``POST /v1/invoices``.

        Each line item is a mapping with ``description``, ``quantity`` and
        ``unitAmount``.
        """
        body = {
            "organizationId": organization_id,
            "customerId": customer_id,
            "lineItems": _normalize_line_items(line_items),
        }
        return self._request("POST", "/v1/invoices", body=body)

    def list(self) -> Any:
        """List invoices. ``GET /v1/invoices``."""
        return self._request("GET", "/v1/invoices")

    def finalize(self, invoice_id: str) -> Any:
        """Finalize a draft invoice. ``POST /v1/invoices/:id/finalize``."""
        return self._request("POST", f"/v1/invoices/{invoice_id}/finalize")

    def pay(self, invoice_id: str) -> Any:
        """Mark an invoice paid. ``POST /v1/invoices/:id/pay``."""
        return self._request("POST", f"/v1/invoices/{invoice_id}/pay")

    def void(self, invoice_id: str) -> Any:
        """Void an invoice. ``POST /v1/invoices/:id/void``."""
        return self._request("POST", f"/v1/invoices/{invoice_id}/void")


class AsyncInvoices(AsyncResource):
    """Asynchronous invoices namespace."""

    async def create(
        self,
        *,
        organization_id: str,
        customer_id: str,
        line_items: Iterable[Mapping[str, Any]],
    ) -> Any:
        """Create a draft invoice. ``POST /v1/invoices``."""
        body = {
            "organizationId": organization_id,
            "customerId": customer_id,
            "lineItems": _normalize_line_items(line_items),
        }
        return await self._request("POST", "/v1/invoices", body=body)

    async def list(self) -> Any:
        """List invoices. ``GET /v1/invoices``."""
        return await self._request("GET", "/v1/invoices")

    async def finalize(self, invoice_id: str) -> Any:
        """Finalize a draft invoice. ``POST /v1/invoices/:id/finalize``."""
        return await self._request("POST", f"/v1/invoices/{invoice_id}/finalize")

    async def pay(self, invoice_id: str) -> Any:
        """Mark an invoice paid. ``POST /v1/invoices/:id/pay``."""
        return await self._request("POST", f"/v1/invoices/{invoice_id}/pay")

    async def void(self, invoice_id: str) -> Any:
        """Void an invoice. ``POST /v1/invoices/:id/void``."""
        return await self._request("POST", f"/v1/invoices/{invoice_id}/void")
